wave_matching/src/icp.cpp

```diff
@@ -75,4 +79,367 @@ bool ICPMatcher::match() {
+
+        auto list = this->icp.correspondences_.get();
+        for (auto it = list->begin(); it != list->end(); ++it) {
+            if (it->index_match > -1) {
+                // it is -1 if there is no match in the target cloud
+                // set up some aliases to make following lines more compact
+                const float &Z1 = (this->target->points[it->index_match].x),
+                            Z2 = (this->target->points[it->index_match].y),
+                            Z3 = (this->target->points[it->index_match].z),
+                            Z4 = (this->ref->points[it->index_query].x),
+                            Z5 = (this->ref->points[it->index_query].y),
+                            Z6 = (this->ref->points[it->index_query].z);
+
+                rg = std::sqrt(Z1*Z1 + Z2*Z2 + Z3*Z3);
+                br = std::atan2(Z2, Z1);
+                az = std::atan(Z3 / std::sqrt(Z1*Z1 + Z2*Z2));
+                j(0,0) = cos(br)*sin(az);
+                j(1,0) = sin(br)*sin(az);
+                j(2,0) = cos(az);
+                j(0,1) = -rg*sin(br)*sin(az);
+                j(1,1) = rg*cos(br)*sin(az);
+                j(0,2) = rg*cos(br)*cos(az);
+                j(1,2) = rg*cos(az)*sin(br);
+                j(2,2) = -rg*sin(az);
+                rg = std::sqrt(Z4*Z4 + Z5*Z5 + Z6*Z6);
+                br = std::atan2(Z5, Z4);
+                az = std::atan(Z6 / std::sqrt(Z4*Z4 + Z5*Z5));
+                j(3,3) = cos(br)*sin(az);
+                j(4,3) = sin(br)*sin(az);
+                j(5,3) = cos(az);
+                j(3,4) = -rg*sin(br)*sin(az);
+                j(4,4) = rg*cos(br)*sin(az);
+                j(3,5) = rg*cos(br)*cos(az);
+                j(4,5) = rg*cos(az)*sin(br);
+                j(5,5) = -rg*sin(az);
+                cov_Z = j * sphere_cov.derived() * j.transpose();
+
+                // coordinate transform jacobian. Order is range, bearing, azimuth for first and 2nd point
+//                [ cos(S2)*sin(S3), -S1*sin(S2)*sin(S3), S1*cos(S2)*cos(S3),               0,                   0,                  0]
+//                [ sin(S2)*sin(S3),  S1*cos(S2)*sin(S3), S1*cos(S3)*sin(S2),               0,                   0,                  0]
+//                [         cos(S3),                   0,        -S1*sin(S3),               0,                   0,                  0]
+//                [               0,                   0,                  0, cos(S5)*sin(S6), -S4*sin(S5)*sin(S6), S4*cos(S5)*cos(S6)]
+//                [               0,                   0,                  0, sin(S5)*sin(S6),  S4*cos(S5)*sin(S6), S4*cos(S6)*sin(S5)]
+//                [               0,                   0,                  0,         cos(S6),                   0,        -S4*sin(S6)]
+
+                // d2J_dx2
+
+                d2J_dX2(0, 0) += 2;
+                d2J_dX2(1, 1) += 2;
+                d2J_dX2(2, 2) += 2;
+
+                d2J_dX2(0, 3) += 2 * Z2 * (sr * sy + cr * cy * sp) +
```

Comment:
consider switching off clang-format by `// clang-format off` for readability, and switch it back on after this giant block of code with `// clang-format on`